### PR TITLE
 :hammer: Speedup anomalist

### PR DESCRIPTION
**Pull request by Marigold**
Speed up loading of UN SDG and GHE datasets. Surprisingly, the bottleneck was loading metadata from dictionary with `dataclasses_json`. This PR replaces it by a custom function that is ~10x faster. This should speed up dataset loading in general, not just for anomalist. It still takes a long time (~1min?) to load these huge datasets (with ~10k variables), but it's manageable when it's run by owidbot.

The new `from_dict` function also raises errors when the type is invalid. This uncovered an inconsistency in our of our old datasets.

It looks like others [have noticed](https://github.com/lidatong/dataclasses-json/issues/228#issuecomment-2355422152) slowness of `dataclasses-json` too. I'll check what makes it so slow on our use case.

Also fixes issues with multidimensional datasets and filters.

Reduce number of GP anomalies - only save anomalies with >= 3 points and p-value < 0.1. That should get rid of useless anomalies like [these from MPI](https://github.com/owid/etl/issues/3436#issuecomment-2464252025).

### CODE REVIEW

File: apps/anomalist/gp_detector.py [hunk 1]
```diff
@@ -118,8 +118,8 @@ def get_score_df(self, df: pd.DataFrame, variable_ids: List[int], variable_mappi
             # Get the data for the current entity and variable
             group = df_wide.loc[(entity_name, variable_id)]
 
-            # Skip if the series has only one or fewer data points
-            if isinstance(group, pd.Series) or len(group) <= 1:
+            # Skip if the series has only three or fewer data points
+            if isinstance(group, pd.Series) or len(group) <= 3:
```
Comment (pabloarosado) [inline apps/anomalist/gp_detector.py:122]: Good idea, maybe this should be applied to all detectors, not just GP (up to you).

File: apps/anomalist/gp_detector.py [hunk 2]
```diff
@@ -160,6 +160,10 @@ def get_score_df(self, df: pd.DataFrame, variable_ids: List[int], variable_mappi
         # Convert z-score into p-value
         df_score_long["p_value"] = 2 * (1 - norm.cdf(np.abs(df_score_long["z"])))
 
+        # Anomalies with p-value < 0.1 are not interesting, drop them. This could be
+        # even stricter
+        df_score_long = df_score_long[df_score_long["p_value"] < 0.1]
```
Comment (pabloarosado) [inline apps/anomalist/gp_detector.py:165]: I'm not sure if this helps much. In the end, we can always filter in the UI. The important thing would be that high scores always mean anomalous points. But if you prefer to hide p>0.1, I don't think it poses any risks of missing out important anomalies.

File: lib/catalog/owid/catalog/processing_log.py [hunk 3]
```diff
@@ -68,6 +68,9 @@ class ProcessingLog(List[LogEntry]):
     # hack for dataclasses_json
     __args__ = (LogEntry,)
 
+    def __init__(self, entries: List[LogEntry | Dict[str, Any]] = []):
+        super().__init__([entry if isinstance(entry, LogEntry) else LogEntry.from_dict(entry) for entry in entries])
```
Comment (pabloarosado) [inline lib/catalog/owid/catalog/processing_log.py:72]: This is a bit obscure to me, but I take your word on this :-D

File: lib/catalog/owid/catalog/utils.py [hunk 4]
```diff
@@ -268,3 +269,55 @@ def hash_any(x: Any) -> int:
...
+        elif dataclasses.is_dataclass(field_type):
+            init_args[field_name] = dataclass_from_dict(field_type, v)  # type: ignore
+        elif isinstance(field_type, type):
+            try:
+                init_args[field_name] = field_type(v)
+            except ValueError as e:
+                # HACK 2024-11-08: we got invalid type into one of the fields, remove it when we fix it
+                # on all staging servers
+                if "invalid literal for int() with base 10: '2020" in str(e):
+                    init_args[field_name] = int(v.replace('"', ""))
```
Comment (pabloarosado) [inline lib/catalog/owid/catalog/utils.py:317]: I'm not sure what's going on here. Maybe we should create an issue to remember to fix this properly?

**Review by pabloarosado — APPROVED** [review]
Thanks @Marigold, I couldn't follow the logic of some parts of the code (maybe it would be good to add a few comments to clarify the intentions), and I haven't tested it myself, but a speed up of 10x when loading variables? Merge this, please! 🎉

File: apps/anomalist/gp_detector.py (hunk 2 above)
Comment (Marigold) [reply apps/anomalist/gp_detector.py:165]: > In the end, we can always filter in the UI

That's right, but it must feel weird for a new user to see GP "anomalies" at the top that are not anomalous at all. I'd keep them if they were useful for "exploration" mode, but I don't think they're useful at all. I'd experiment with filtering them and if we miss them in the future, we can just put them back.

File: lib/catalog/owid/catalog/processing_log.py (hunk 3 above)
Comment (Marigold) [reply lib/catalog/owid/catalog/processing_log.py:72]: It accepts both `LogEntry` and `dict` when initializing. Clarified that in a comment.

File: lib/catalog/owid/catalog/utils.py (hunk 4 above)
Comment (Marigold) [reply lib/catalog/owid/catalog/utils.py:317]: It comes [from a typo](https://github.com/owid/etl/pull/3517/files#diff-00e51537a22bb68140652296351986cdc3e1d430d97a27b103c4b83b62eb4280) in one old metadata file. I'll remove it the next time we increase ETL_EPOCH.

**Comment by Marigold** [comment]
After more experimentation, I've found that adding [caching to dataclasses-json](https://github.com/lidatong/dataclasses-json/issues/228#issuecomment-2355422152) would get it almost on par with our custom `from_dict` method. I've also tried https://github.com/cakemanny/fastclasses-json which was fast (again, on par with our custom method), but I was getting tons type hints errors from pylance.

I'll leave a comment about revisiting this and possibly using third-party libraries in the future.

File: apps/anomalist/gp_detector.py (hunk 1 above)
Comment (Marigold) [reply apps/anomalist/gp_detector.py:122]: I'll leave this decision up to you. GP is not useful with just 3 points, but other detectors might be.

**Pull request merged**